CI/CD Docs Updates (#7563)
RasaHQ/rasa: docs/docs/setting-up-ci-cd.mdx

```yaml
jobs:
  training-testing:
    name: Training and Testing
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v1
      - name: Rasa Train and Test GitHub Action
        uses: RasaHQ/rasa-train-test-gha@main
        with:
          requirements_file: requirements.txt
          test_type: all
          cross_validation: true
          publish_summary: true
          github_token: ${{ secrets.GITHUB_TOKEN }}
      - name: Upload model
        if: github.ref == 'refs/heads/master'
        uses: actions/upload-artifact@master
        with:
          name: model
          path: models
```

In this pipeline, the Rasa Train-Test Github Action is performing data validation, model training, and story testing
in the first step and the model file is uploaded as an artifact in the second step.

The complete list of configurable parameters for the Rasa Train-Test Github Action is available in the repository's
[README](https://github.com/RasaHQ/rasa-train-test-gha#input-arguments).

When `publish_summary` is set to `true`, this action will automatically publish the model's test results to the associated
Pull Request as a comment:

> <img alt="image" src={useBaseUrl("/img/train-test-github-action.png")} />

Reviewer: ```suggestion
<img alt="image" src={useBaseUrl("/img/train-test-github-action.png")} />

After the pull request is opened, tag a reviewer from your team to see the results. The reviewer can then approve the change if the results from the tests are positive, allowing you to continue to Continuous Deployment.
```

Reviewer: @b-quachtran looks good, just added a comment to wrap it up at the end.

Author: I would err on the side of being less specific here about tagging and reviewing since it's dependent on their individual review processes. Wdyt about this instead:

```suggestion
<img alt="image" src={useBaseUrl("/img/train-test-github-action.png")} />

The pull request can be approved or denied based on the evaluation results and, in many cases, you will want to automate the model's deployment if all CI checks pass. You can continue to the next section to learn more about Continuous Deployment. 
```

Reviewer: looks great!

other indicator is present. For example, if your code is hosted on Github,
you can make a test run only if the pull request has a certain label (e.g. “NLU testing required”).

### CI Pipeline Overview

> There are different approaches to applying a CI pipeline depending on the number of changes.

Author: ```suggestion
Your CI pipeline should include model training and testing as steps to streamline the deployment process. 
```

> All follow a mix of training the new model, testing to compare to older versions,

Author: ```suggestion
```

> and setting up an easy CD pipeline to upload the new model.

Author: ```suggestion
```

> The first step after saving new training data is to kick off the pipeline. (Is it usually kicked off

Author: It should be kicked off automatically in the pipeline where it does a diff on your training data & domain file

Reviewer 2: Something like: `This can be initiated manually or when you create or update a pull request`

Reviewer: adding greg's suggestion

manually or triggered by an event like saving a new file?)
(need different pipelines / options based on how big the changes are - also should we lay out how

> small or big of changes warrant which tests?) Typically this is done by xyz

Author: I think we can leave out how this is done since it varies based on the pipeline and it's a standard task

Author: Generally tests will be the same regardless of how big the changes are. It's more of a binary thing where if you need to train a new model, you'll also need to run tests.

Reviewer 2: Agree that we don't need to address the size of the change.

> Next, you need to pick what level of tests to run based on the amount of data changed. The various types

Author: Testing will generally be the same if they need to be run

> Next, you need to pick what level of tests to run based on the amount of data changed. The various types

Author: ```suggestion
```

are laid out in [Testing Your Assistant](https://rasa.com/docs/rasa/testing-your-assistant), but a good

> rule of thumb is to always run your test stories to ensure no bugs were introduced. Larger changes

Author: This can be re-worded to mention running data validation, NLU cross validation, and story testing, and then referring to the Testing Your Assistant page if they want more details

may need a full NLU test with cross validation to look out for confusion between intents.

> After testing and posting your results for your team to see, have a person or team of reviewers

Author: I'd take this part about reviewing PR's out

Reviewer 2: I agree, let's not worry about roles right now.

to check the test results with your developer. An easy process to help this is to have your CI

> pipeline automatically create a pull request with the test results in a comment. That way, the

Author: A developer will have to manually create the PR but the test results will get posted automatically

Reviewer 2: Something like `The tests results from your CI pipeline should automatically be added as a comment or artifact of the PR.`

reviewer can see the changes along with the improved results.

> Once the reviewer merges you can then initiate your CD pipeline to add the new model

Author: ```suggestion
Once the new model is trained and tested, it can be deployed automatically using a Continuous Deployment pipeline. 
```

> to another environment.

Author: ```suggestion
```

other indicator is present. For example, if your code is hosted on Github,
you can make a test run only if the pull request has a certain label (e.g. “NLU testing required”).

### CI Pipeline Overview

Your CI pipeline should include model training and testing as steps to streamline the deployment process. 
The first step after saving new training data is to kick off the pipeline. This can be initiated manually 
or when you create or update a pull request.

Next, you need to run various sets of test to see the impact of your changes. This includes running 
tests for data validation, NLU cross validation, and story testing. For 

> more information about testing, see [Testing Your Assistant](https://rasa.com/docs/rasa/testing-your-assistant). 

Author: ```suggestion
more information about testing, see [Testing Your Assistant](./testing-your-assistant.mdx). 
```

may need a full NLU test with cross validation to look out for confusion between intents.

After testing and posting your results for your team to see, have a person or team of reviewers
to check the test results with your developer. An easy process to help this is to have your CI
pipeline automatically create a pull request with the test results in a comment. That way, the
reviewer can see the changes along with the improved results.

Once the reviewer merges you can then initiate your CD pipeline to add the new model
to another environment.

### GitHub Actions CI Pipeline

You can use the [Rasa Train-Test Github Action](https://github.com/RasaHQ/rasa-train-test-gha)
in your CI pipeline to automatically perform data validation, training, and testing.

An example CI pipeline using the Github Action is shown below:

```yaml
jobs:
  training-testing:
    name: Training and Testing
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v1
      - name: Rasa Train and Test GitHub Action
        uses: RasaHQ/rasa-train-test-gha@main
        with:
          requirements_file: requirements.txt

>           test_type: all

Reviewer 2: I think we should add `data_validate: true` and `rasa_train:true` before this line. Even though though the default is true.

Reviewer: @b-quachtran can you take this one?

Author: Yep ✅
